src/dmd/dtemplate.d

```diff
@@ -6244,12 +6244,27 @@ extern (C++) class TemplateInstance : ScopeDsymbol
             // 2. become non-speculative if a sibling isn't speculative
 
             // Determine necessity of tinst before tnext.
-            if (tinst && tinst.needsCodegen())
+            if (tinst)
             {
-                minst = tinst.minst; // cache result
-                assert(minst);
-                assert(minst.isRoot() || minst.rootImports());
-                return true;
+                const needsCodegen = tinst.needsCodegen();
+                if (global.params.allInst)
+                {
+                    if (needsCodegen)
+                    {
+                        minst = tinst.minst; // cache result
+                        assert(minst);
+                        assert(minst.isRoot());
+                        return true;
+                    }
+                }
+                else
+                {
+                    if (tinst.minst) // not speculative
+                    {
+                        minst = tinst.minst; // cache result
+                        return needsCodegen;
+                    }
```

Comment:
For non-`allinst`, this is now consistent with https://github.com/dlang/dmd/blob/4c3701de013f1cb3c3eb31c4207933c26fca9f9c/src/dmd/dtemplate.d#L6315-L6320 for non-speculative children.